storage/innobase/log/log0online.cc

```diff
@@ -273,6 +278,83 @@ class log_parse_buffer final : public log_buffer<RECV_PARSING_BUF_SIZE> {
 
   MY_NODISCARD lsn_t get_end_lsn() const noexcept { return end_lsn; }
 
+  void set_checkpoint_lsn(lsn_t checkpoint_lsn_) noexcept {
+    ut_ad(mutex_own(&log_bmp_sys_mutex));
+    ut_ad(checkpoint_lsn_ >= checkpoint_lsn);
+    checkpoint_lsn = checkpoint_lsn_;
+    parse_start_lsn = 0;
+    parse_stop_lsn = 0;
+    is_checkpoint_block_reached = false;
+  }
+
+  lsn_t get_parse_start_lsn() const noexcept { return parse_start_lsn; }
+
+  lsn_t set_parse_start_lsn(lsn_t lsn) noexcept {
+    return parse_start_lsn = lsn;
+  }
+
+  lsn_t get_parse_stop_lsn() const noexcept { return parse_stop_lsn; }
+
+  void stop_parsing_at(lsn_t lsn) noexcept {
+    ut_ad(lsn >= checkpoint_lsn && lsn >= current_lsn);
+    parse_stop_lsn = lsn;
+    if (is_checkpoint_block_reached && parse_stop_lsn >= checkpoint_lsn) {
+      move_unprocessed_to_front(parse_result::PAST_END);
+    }
+  }
+
+  bool checkpoint_reached() const noexcept {
+    return is_checkpoint_block_reached;
+  }
+
+  void reset() noexcept {
+    parse_status = parse_result::OK;
+    current_ptr = buffer.cbegin();
+    end_ptr = buffer.begin();
+    current_size = 0;
+
+    start_lsn = 0;
+    current_lsn = 0;
+    checkpoint_lsn = 0;
+    end_lsn = 0;
+    parse_start_lsn = 0;
+    parse_stop_lsn = 0;
+    is_checkpoint_block_reached = false;
+  }
+
+  void check_update_parse_stop_lsn(log_read_buffer &from,
+                                   log_buffer::size_type data_len) noexcept {
+    auto read_buf_current_lsn = from.get_current_lsn();
+
+    if (!is_checkpoint_block_reached &&
+        checkpoint_lsn <= read_buf_current_lsn + OS_FILE_LOG_BLOCK_SIZE) {
+      // The checkpoint_lsn points to this block. Starting from this block we
+      // are looking for a beginning of the next log record group which starts
+      // after checkpoint_lsn. Parsing should stop at that point.
+      is_checkpoint_block_reached = true;
+    }
+
+    if (is_checkpoint_block_reached && parse_stop_lsn == 0) {
+      if (data_len < OS_FILE_LOG_BLOCK_SIZE) {
+        // Block is filled in partially. Parsing should stop at the end of the
+        // data block.
+        parse_stop_lsn = read_buf_current_lsn + data_len;
+      } else {
+        auto rec_group_offset = log_block_get_first_rec_group(from.ccurrent());
+        if (rec_group_offset > 0 &&
+            read_buf_current_lsn + rec_group_offset >= checkpoint_lsn) {
```

Comment:
If rec_group_offset is 0, next record group starts in one of the next blocks. It will parse current block. Read next one and most probably parse_stop_lsn will be found there while actually parsing data.

> the parsing_stop_len is < checkpoint_lsn
Is this one from recovery code?

Reply:
Yes, in recovery code recv_scan_log_recs().

      if (recv_sys->parse_start_lsn < recv_sys->checkpoint_lsn) {
        /* We start to parse log records even before
        checkpoint_lsn, from the beginning of the log
        block which contains the checkpoint_lsn.

        That's because the first group of log records
        in the log block, starts before checkpoint_lsn,
        and checkpoint_lsn could potentially point to
        the middle of some log record. We need to find
        the first group of log records that starts at
        or after checkpoint_lsn. This could be only
        achieved by traversing all groups of log records
        that start within the log block since the first
        one (to discover their beginnings we need to
        parse them). However, we don't want to report
        missing tablespaces for space_id in log records
        before checkpoint_lsn. Hence we need to ignore
        those records and that's why we need a counter
        of bytes to ignore. */

Reply:
As I understand they start parsing at `recv_sys->parse_start_lsn = scanned_lsn + log_block_get_first_rec_group(log_block)` but still skip data until first record group starting after checkpoint_lsn is found.
In my change I do similar thing but in a bit different way. In `log_online_follow_log_seg()` I calculate LSN for first record group in block

`const auto parse_start_lsn = log_bmp_sys->read_buf.get_current_lsn() + first_rec_group_offset;`

In case it points before checkpoint, log parsing still begins from this point (data from flock start till this offset is skipped) but parse_start_lsn isn't stored in parse buffer. It will be calculated while parsing the log.

Reply:
why not use log_block_get_first_rec_group(block) ? what is difference between log_block_get_first_rec_group(block) vs first_rec_group_offset?

Reply:
In this method (log_block_get_first_rec_group) the log_block_get_first_rec_group() is actually used. Maybe you refer to some different place in code?